### PR TITLE
RISC-V: implement calls without trampolines

### PR DESCRIPTION
**Pull request by janvrany**
This PR adds support for calls without trampolines as a temporary solution until #2017 is implemented.

### CODE REVIEW

**Comment by janvrany** [comment]
As of now, there's a bug somewhere in linkage making a number  `LinkageTest`s failing - I decided to keep this PR as draft until I hunt down that bug and fix it so tests pass. 

Commit 28f7225 is the one that add support for calls, the rest is just a bunch of fixes and cleanups. These should probably go into separate PR as they have nothing to do with calls, but I'll split this PR once I have tests passing (to avoid "oops, fix this" commits). I'd appreciate pre-review of commits before 28f7225, especially commit 31b8e5a (I'm not 100% sure of it but clearly it prevents `TR_ASSERT()`ion failure when running `LinkageTests`.

Finally, @knn-k, maybe it'd make sense to do d33e1ad also for AArch64?

**Comment by janvrany** [comment]
jenkins build all

**Comment by knn-k** [comment]
It is not only AArch64 but p/z/arm platforms also use the name `_numberOfDependencyGPRegisters`.  An example:
https://github.com/eclipse/omr/blob/cd2fd54f0bf5f04086c7481490b090ec68c03b4c/compiler/p/codegen/OMRLinkage.hpp#L121

I agree that the name is confusing.

**Comment by janvrany** [comment]
Commit df0db05 fixes passing of integer parameters on stack. 

However, passing float arguments is still broken. The problem is that on RISC-V, if FP there's no more FP argument register available (that is, for 9th and more FP argument), it is passed in integer argument register (if available). An excerpt from
[RISC-V PSABI spec](https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#hardware-floating-point-calling-convention):

> A real floating-point argument is passed in a floating-point argument register if it is no more than FLEN bits wide and at least one floating-point argument register is available. Otherwise, it is passed according to the integer calling convention. When a floating-point argument narrower than FLEN bits is passed in a floating-point register, it is 1-extended (NaN-boxed) to FLEN bits.

Funny enough, jit-to-native FP tests pass even for 9 arguments because jitted code does not clobber `a0` (first integer argument register). 

I'll update this PR once correct behavior is implemented.

**Comment by janvrany** [comment]
jenkins build riscv

**Comment by janvrany** [comment]
I have fixed all remaining issues so now both `CallTest` and `LinkageTest`. Therefore I think this PR is ready for review.

File: compiler/riscv/codegen/OMRLinkage.hpp [hunk 1]
```diff
@@ -340,12 +340,13 @@ class OMR_EXTENSIBLE Linkage : public OMR::Linkage
...
     * @brief Returns a MemoryReference for an outgoing argument
     * @param[in] argMemReg : register pointing to address for the outgoing argument
+    * @param[in] offset : offset from argMemReg in bytes
     * @param[in] argReg : register for the argument
     * @param[in] opCode : instruction OpCode for store to memory
     * @param[out] memArg : struct holding memory argument information
     * @return MemoryReference for the argument
     */
-   virtual TR::MemoryReference *getOutgoingArgumentMemRef(TR::Register *argMemReg, TR::Register *argReg, TR::InstOpCode::Mnemonic opCode, TR::RVMemoryArgument &memArg);
+   virtual TR::MemoryReference *getOutgoingArgumentMemRef(TR::Register *argMemReg, int offset, TR::Register *argReg, TR::InstOpCode::Mnemonic opCode, TR::RVMemoryArgument &memArg);
```
Comment (0xdaryl) [inline compiler/riscv/codegen/OMRLinkage.hpp:349]: The convention in the compiler is to use an explicit-sized integer type:  `int32_t`

File: compiler/riscv/codegen/OMRMachine.cpp [hunk 2]
```diff
@@ -465,14 +465,18 @@ static void registerCopy(TR::Instruction *precedingInstruction,
                          TR::RealRegister *sourceReg,
                          TR::CodeGenerator *cg)
    {
+   TR_ASSERT(sourceReg->getKind() == rk, "Source register kind mismatch.");
+   TR_ASSERT(sourceReg->getKind() == targetReg->getKind(), "Source and target register kind mismatch.");
```
Comment (0xdaryl) [inline compiler/riscv/codegen/OMRMachine.cpp:469]: These should be `TR_ASSERT_FATAL` to be of any use.

File: compiler/riscv/codegen/OMRMachine.cpp [hunk 3]
```diff
@@ -465,14 +465,18 @@ static void registerCopy(TR::Instruction *precedingInstruction,
...
+
    TR::Node *node = precedingInstruction->getNode();
    switch (rk)
       {
       case TR_GPR:
          generateITYPE(TR::InstOpCode::_addi, node, targetReg, sourceReg, 0, cg, precedingInstruction);
          break;
       case TR_FPR:
-         TR_ASSERT(false, "Unsupported RegisterKind.");
+         TR_ASSERT(sourceReg->isSinglePrecision() == targetReg->isSinglePrecision(), "Source and target register size mismatch");
```
Comment (0xdaryl) [inline compiler/riscv/codegen/OMRMachine.cpp:478]: `TR_ASSERT_FATAL`

File: compiler/riscv/codegen/RVInstruction.cpp [hunk 4]
```diff
@@ -495,15 +495,15 @@ uint8_t *TR::JtypeInstruction::generateBinaryEncoding() {
          }
       else
          {
-         TR_ASSERT(0, "Non-recursive calls not (yet) supported");
+         offset = (uint8_t*)getSymbolReference()->getMethodAddress() - cursor;
```
Comment (0xdaryl) [inline compiler/riscv/codegen/RVInstruction.cpp:498]: Please make this a `reinterpret_cast` as the compiler is starting to adopt the C++ casts exclusively.

File: compiler/riscv/codegen/RVInstruction.cpp [hunk 5]
```diff
@@ -495,15 +495,15 @@ uint8_t *TR::JtypeInstruction::generateBinaryEncoding() {
...
          }
       }
    else
       {
-      uintptr_t destination = (uintptr_t)(getLabelSymbol()->getCodeLocation());
+      uint8_t* destination = (uint8_t*)(getLabelSymbol()->getCodeLocation());
```
Comment (0xdaryl) [inline compiler/riscv/codegen/RVInstruction.cpp:503]: Please make this a `reinterpret_cast` as the compiler is starting to adopt the C++ casts exclusively.

File: compiler/riscv/codegen/RVSystemLinkage.cpp [hunk 6]
```diff
@@ -686,36 +717,49 @@ int32_t TR::RVSystemLinkage::buildArgs(TR::Node *callNode,
...
                }
             numFloatArgs++;
             break;
          } // end of switch
       } // end of for
 
-   // NULL deps for non-preserved and non-system regs
-   while (numIntegerArgs < properties.getNumIntArgRegs())
+   // NULL deps for unused integer argument registers
+   for (int i = numIntegerArgs + numFloatArgsPassedInGPRs; i < properties.getNumIntArgRegs(); i++)
```
Comment (0xdaryl) [inline compiler/riscv/codegen/RVSystemLinkage.cpp:729]: Compiler convention is to use explicitly-sized integers: `int32_t`

File: fvtest/compilertriltest/CallTest.cpp [hunk 7]
```diff
@@ -27,17 +27,22 @@ class CallTest : public TRTest::JitTest {};
 int32_t oracleBoracle(int32_t x) { return x + 1123; } // Randomish number to avoid accidental test passes.
 
 TEST_F(CallTest, icallOracle) { 
+
+	SKIP_ON_PPC(MissingImplementation) << "Test is skipped on POWER because calls are not currently supported (see issue #1645)";
```
Comment (0xdaryl) [inline fvtest/compilertriltest/CallTest.cpp:31]: Fix indenting.

File: compiler/riscv/codegen/RVSystemLinkage.cpp [hunk 8]
```diff
@@ -774,9 +818,35 @@ TR::Register *TR::RVSystemLinkage::buildDispatch(TR::Node *callNode)
...
+          * more often than not, non-jitted functions (library / runtime functions) are too far to fit in
+          * the immediate value.
+          *
+          * So, until a trampolines are implemented, we pay the price and load the target address into
+          * register.
+          *
+          * Note, that here we load the target address into link register (`ra`). It's going to be clobbered
+          * anyways and this way we do not need to allocate another one.
+          */
+         loadConstant64(cg(), callNode, (int64_t)targetAddr, ra);
```
Comment (0xdaryl) [inline compiler/riscv/codegen/RVSystemLinkage.cpp:847]: Please make this a `static_cast` as the compiler is starting to adopt the C++ casts exclusively.

**Review by 0xdaryl — CHANGES_REQUESTED** [review]

**Comment by janvrany** [comment]
jenkins build riscv

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**